{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.output_wrapper, .output {    height:auto !important;max-height:1000px;  /* your desired max-height here */}.output_scroll {    box-shadow:none !important;    webkit-box-shadow:none !important;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# modules\n",
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.output_wrapper, .output {    height:auto !important;max-height:1000px;  /* your desired max-height here */}.output_scroll {    box-shadow:none !important;    webkit-box-shadow:none !important;}</style>\"))\n",
    "\n",
    "import random\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "import argparse\n",
    "import subprocess\n",
    "from time import clock\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from itertools import groupby\n",
    "import itertools\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Alphabet import IUPAC, Gapped, generic_dna\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from sys import argv\n",
    "from Bio import AlignIO, SeqIO, Phylo\n",
    "from time import time\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import seaborn as sns; sns.set()\n",
    "from scipy import stats\n",
    "from random import randint,randrange\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--Ne: 10000\n",
      "--number_of_generations: 10000\n",
      "--percent_of_recombinating_sp: 0\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 10000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-3d2c659630f2>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    184\u001b[0m             \u001b[0mstep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0man_elementary_act_of_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNe\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpercent_of_recombinating_sp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"nuc\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"first\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumber_of_generations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 186\u001b[0;31m                 \u001b[0mstep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0man_elementary_act_of_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNe\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpercent_of_recombinating_sp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"nuc\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    187\u001b[0m             \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclock\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-3d2c659630f2>\u001b[0m in \u001b[0;36man_elementary_act_of_evolution\u001b[0;34m(Ne, seq_len, percent_of_recombinating_sp, mutations_per_gen, nuc_mode, population_mode)\u001b[0m\n\u001b[1;32m    140\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmultiply_seqs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mselected_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"crossingover_it\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmultiply_seqs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-3d2c659630f2>\u001b[0m in \u001b[0;36mmultiply_seqs\u001b[0;34m(result, work_list, mutations_per_gen, mode)\u001b[0m\n\u001b[1;32m    134\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 136\u001b[0;31m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    137\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-3d2c659630f2>\u001b[0m in \u001b[0;36mmutate\u001b[0;34m(seq, mutations_per_gen, mode)\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0;31m# creating mutation function -- ok\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m         \u001b[0mseq_cp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"new_population\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#%%notify\n",
    "\n",
    "\n",
    "def status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations):\n",
    "    print(\"Recieved parameters: \\n \\t Ne: {Ne}, \\n \\t seq_len: {seq_len}, \\n \\t percent_of_recombinating_sp: {percent_of_recombinating_sp}, \\n \\t mutations_per_gen: {mutations_per_gen}, \\n \\t number_of_generations: {number_of_generations}\".format(Ne = Ne,\n",
    "seq_len = seq_len,\n",
    "percent_of_recombinating_sp = percent_of_recombinating_sp,\n",
    "mutations_per_gen = mutations_per_gen,\n",
    "number_of_generations = number_of_generations))\n",
    "\n",
    "#status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)  \n",
    "\n",
    "\n",
    "\n",
    "# creating a model\n",
    "\n",
    "def an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,nuc_mode=None,population_mode=None):\n",
    "    random.seed(666)\n",
    "    #status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)\n",
    "    \n",
    "    if nuc_mode == \"nuc\":  # выбираем, с чем работаем\n",
    "        seed = \"ATGC\"\n",
    "    else:\n",
    "        seed = \"1234\"\n",
    "    \n",
    "    # creating mutation function -- ok\n",
    "    def mutate(seq, mutations_per_gen, mode=None):\n",
    "        seq_cp = seq.copy()\n",
    "\n",
    "        if mode == \"new_population\":\n",
    "            mutations_per_gen_tmp = mutations_per_gen + int(0.01 * seq_len)\n",
    "            for i in range(mutations_per_gen_tmp):\n",
    "                mutation_site = randrange(len(seq_cp))\n",
    "                seq_cp[mutation_site] = random.choice(seed)     \n",
    "            return seq_cp \n",
    "        else:\n",
    "            r = random.uniform(0, 1) # чтобы было меньше мутаций\n",
    "            if r > 1/10:\n",
    "                return seq_cp\n",
    "            else:\n",
    "                mutations_per_gen_tmp = mutations_per_gen\n",
    "                for i in range(mutations_per_gen_tmp):\n",
    "                    mutation_site = randrange(len(seq_cp))\n",
    "                    seq_cp[mutation_site] = random.choice(seed)     \n",
    "                return seq_cp\n",
    "         \n",
    "    if population_mode == \"first\":   # --ok\n",
    "        # creating seqs\n",
    "        seq = [random.choice(seed) for i in range(seq_len)]\n",
    "        population = [mutate(seq, mutations_per_gen, \"new_population\") for x in range(Ne)]\n",
    "    else:\n",
    "        population = population_mode\n",
    "        Ne = len(population)\n",
    "    # breeding\n",
    "    selected_list = []\n",
    "    number_of_species = int(percent_of_recombinating_sp * len(population))\n",
    "    if number_of_species%2 != 0:\n",
    "        number_of_species -= 1\n",
    "        \n",
    "    for i in range(number_of_species): # --ok\n",
    "        p = randrange(len(population))\n",
    "        selected_list.append(population.pop(p))\n",
    "    \n",
    "    # crossingover\n",
    "    def crossingover(seq_in1, seq_in2):\n",
    "        number_of_parts = random.randrange(0, 2, 1)\n",
    "        if number_of_parts > 0:   \n",
    "            shuffled_1 = []\n",
    "            shuffled_2 = []\n",
    "            left = 0\n",
    "            right = len(seq_in1)\n",
    "            ind = True\n",
    "            for i in range(number_of_parts):\n",
    "                n = number_of_parts - i\n",
    "                #print(right - left)\n",
    "                if right - left > n + 1:\n",
    "                    #print(i, left, right, ind)\n",
    "                    \n",
    "                    #print(\"right//n\", right//n)\n",
    "                    position = random.randint(left, right//n)\n",
    "                    #print(\"position\", position)\n",
    "                    if ind == True:\n",
    "                        shuffled_1.append(seq_in1[left:position])\n",
    "                        shuffled_2.append(seq_in2[left:position])\n",
    "                        #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "                        ind = False\n",
    "                        left = position\n",
    "                        continue\n",
    "                    if ind == False:\n",
    "                        shuffled_1.append(seq_in2[left:position])\n",
    "                        shuffled_2.append(seq_in1[left:position])\n",
    "                        #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "                        ind = True\n",
    "                        left = position\n",
    "                        continue\n",
    "                else:\n",
    "                    #print(\"out\")\n",
    "                    if ind == True:\n",
    "                        shuffled_1.append(seq_in1[left:position])\n",
    "                        shuffled_2.append(seq_in2[left:position])\n",
    "                        #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "                    if ind == False:\n",
    "                        shuffled_1.append(seq_in2[left:position])\n",
    "                        shuffled_2.append(seq_in1[left:position])\n",
    "                        #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "                    return shuffled_1, shuffled_2                        \n",
    "            if ind == True:\n",
    "                shuffled_1.append(seq_in1[left:])\n",
    "                shuffled_2.append(seq_in2[left:])\n",
    "                #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "            if ind == False:\n",
    "                shuffled_1.append(seq_in2[left:])\n",
    "                shuffled_2.append(seq_in1[left:])\n",
    "                #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "            return shuffled_1, shuffled_2\n",
    "        else:\n",
    "            return seq_in1, seq_in2\n",
    "\n",
    "    \n",
    "    \n",
    "    # multipy selected seqs\n",
    "    def multiply_seqs(result, work_list, mutations_per_gen, mode=None):\n",
    "        for i, j in zip(work_list[0::2], work_list[1::2]):\n",
    "            if mode == \"crossingover_it\":\n",
    "                \n",
    "                p, q = crossingover(i, j)\n",
    "                \n",
    "                p = list(itertools.chain.from_iterable(p))\n",
    "                q = list(itertools.chain.from_iterable(q))\n",
    "\n",
    "                for k in range(1): \n",
    "                    result.append(mutate(p, mutations_per_gen))\n",
    "                    result.append(mutate(q, mutations_per_gen))\n",
    "            else:\n",
    "                for k in range(1):\n",
    "                    result.append(mutate(i, mutations_per_gen))\n",
    "                    result.append(mutate(j, mutations_per_gen))\n",
    "        return result\n",
    "    \n",
    "    result = []\n",
    "    result = multiply_seqs(result, selected_list, mutations_per_gen, \"crossingover_it\")\n",
    "    result = multiply_seqs(result, population, mutations_per_gen)\n",
    "    \n",
    "    result = random.sample(result, Ne)\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "def decoder(result):\n",
    "    def make_it_in_nucl(seq):\n",
    "        dic = {1:\"A\",2:\"T\",3:\"G\",4:\"C\"}\n",
    "        for j in range(len(seq)):\n",
    "            seq[j] = dic[int(seq[j])]\n",
    "        return seq\n",
    "    for i in range(len(result)):\n",
    "        result[i] = make_it_in_nucl(result[i])\n",
    "    return result\n",
    "\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/Simulation\"\n",
    "outputdir = os.path.abspath(root + '/result_0.01')\n",
    "    \n",
    "if not os.path.exists(outputdir):\n",
    "    os.makedirs(outputdir)\n",
    "\n",
    "    \n",
    "# loading parameters\n",
    "mutations_per_gen = 1\n",
    "seq_len = 1000\n",
    "\n",
    "\n",
    "os.chdir(outputdir)\n",
    "for Ne in [x * 10000 for x in range(1, 10, 2)]: #+ [x * 1000 for x in range(2, 10, 2)]:\n",
    "    print(\"--Ne:\", Ne)\n",
    "    for number_of_generations in [x * 10000 for x in range(1, 10, 2)]:# + [x * 1000 for x in range(2, 10, 2)]:\n",
    "        print(\"--number_of_generations:\", number_of_generations)\n",
    "        for percent_of_recombinating_sp in [0, 0.01, 0.02, 0.05, 0.1, 0.2, 0.5, 1]:\n",
    "            print(\"--percent_of_recombinating_sp:\", percent_of_recombinating_sp)\n",
    "            print(\" \")\n",
    "            \n",
    "            status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)  \n",
    "            \n",
    "            start = clock()\n",
    "            step = an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,\"nuc\",\"first\")\n",
    "            for i in range(number_of_generations):\n",
    "                step = an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,\"nuc\",step)\n",
    "            end = clock()\n",
    "            \n",
    "            step = random.sample(step, 100) \n",
    "            \n",
    "        \n",
    "            def count_dict(mystring):\n",
    "                d = {}\n",
    "                for w in mystring: \n",
    "                    d[w] = mystring.count(w)\n",
    "                #print(d.values(), min(d.values())/len(mystring))\n",
    "                if min(d.values())/len(mystring) < 0.25:\n",
    "                    return -1\n",
    "                else:\n",
    "                    return 1\n",
    "\n",
    "            def slice_maker(step):\n",
    "                for i in range(len(step[0])):\n",
    "                    string = \"\"\n",
    "                    for j in range(len(step)):\n",
    "                        string += step[j][i]\n",
    "                    #print(string)\n",
    "                    if count_dict(string) == -1:\n",
    "                        for j in range(len(step)):\n",
    "                            step[j][i] = \"N\"\n",
    "                    \n",
    "            slice_maker(step)\n",
    "\n",
    "            #step = decoder(step)\n",
    "            #print(step)\n",
    "            \n",
    "            \n",
    "            align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "            for i in range(len(step)):\n",
    "                align.add_sequence(str(i), \"\".join(step[i]))\n",
    "            \n",
    "            SeqIO.write(align, outputdir + \"/\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \"_align.fasta\", \"fasta\")\n",
    "        \n",
    "            print(\"timing is\", round(end - start,5), \"seconds\")\n",
    "            \n",
    "            try:\n",
    "                line1 = 'snp-sites -mvp -o {out_file} {in_file}'.format(out_file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len),\n",
    "                    in_file = outputdir + \"/\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \"_align.fasta\")\n",
    "                p = subprocess.Popen(line1, shell=True)\n",
    "                p.wait()\n",
    "            \n",
    "                line2 = 'vcftools --hap-r2 --vcf {in_file} --out {out_file}'.format(in_file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".vcf\", \n",
    "                        out_file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len))\n",
    "                p = subprocess.Popen(line2, shell=True)\n",
    "                p.wait()\n",
    "                \n",
    "                    \n",
    "                df = pd.read_csv(outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".hap.ld\", sep='\\t',index_col=False)\n",
    "                df[\"Len\"] = df[\"POS2\"] - df[\"POS1\"]\n",
    "                \n",
    "                df_name = outputdir + \"/LD_df_\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".csv\"\n",
    "                df.to_csv(path_or_buf = df_name)\n",
    "                \n",
    "                line3 = 'Rscript \"/home/pavel/Desktop/Work/Project/Simulation/R_plot.R\" --no-save --no-restore --args {outputdir} {plot_name} 0.0 {df_name}'.format(outputdir = outputdir, \n",
    "                plot_name = outputdir + \"/Ne-\" + str(Ne) + \"_gen_sp-\" + str(number_of_generations) + \"_rec-\" + str(percent_of_recombinating_sp) + \"_mut-\" + str(mutations_per_gen) + \"_seq_l-\" + str(seq_len),\n",
    "                df_name = df_name)\n",
    "                p = subprocess.Popen(line3, shell=True)\n",
    "                p.wait()\n",
    "                \n",
    "                            \n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".phylip\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".snp_sites.aln\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".vcf\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len))\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "                    \n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".log\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".hap.ld\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "                    \n",
    "                    \n",
    "            except ZeroDivisionError:\n",
    "                continue\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--Ne: 5000\n",
      "--number_of_generations: 10000\n",
      "--percent_of_recombinating_sp: 0.001\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.001, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 687.60453 seconds\n",
      "--percent_of_recombinating_sp: 0.005\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.005, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 727.65623 seconds\n",
      "--percent_of_recombinating_sp: 0.01\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.01, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 803.02357 seconds\n",
      "--percent_of_recombinating_sp: 0.02\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.02, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 768.75202 seconds\n",
      "--percent_of_recombinating_sp: 0.05\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.05, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 815.20157 seconds\n",
      "--percent_of_recombinating_sp: 0.1\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.1, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 891.9209 seconds\n",
      "--percent_of_recombinating_sp: 0.2\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.2, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 932.76693 seconds\n",
      "--percent_of_recombinating_sp: 0.5\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.5, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 1502.03704 seconds\n",
      "--percent_of_recombinating_sp: 1\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 1, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 10000\n",
      "timing is 2111.81278 seconds\n",
      "--number_of_generations: 30000\n",
      "--percent_of_recombinating_sp: 0.001\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.001, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2352.88628 seconds\n",
      "--percent_of_recombinating_sp: 0.005\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.005, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2112.49211 seconds\n",
      "--percent_of_recombinating_sp: 0.01\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.01, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2046.60899 seconds\n",
      "--percent_of_recombinating_sp: 0.02\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.02, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2081.08499 seconds\n",
      "--percent_of_recombinating_sp: 0.05\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.05, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2183.80886 seconds\n",
      "--percent_of_recombinating_sp: 0.1\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.1, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2366.55589 seconds\n",
      "--percent_of_recombinating_sp: 0.2\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.2, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 2728.18727 seconds\n",
      "--percent_of_recombinating_sp: 0.5\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.5, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 3931.14877 seconds\n",
      "--percent_of_recombinating_sp: 1\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 1, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 30000\n",
      "timing is 5843.89932 seconds\n",
      "--number_of_generations: 50000\n",
      "--percent_of_recombinating_sp: 0.001\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.001, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 50000\n",
      "timing is 3388.46121 seconds\n",
      "--percent_of_recombinating_sp: 0.005\n",
      " \n",
      "Recieved parameters: \n",
      " \t Ne: 5000, \n",
      " \t seq_len: 1000, \n",
      " \t percent_of_recombinating_sp: 0.005, \n",
      " \t mutations_per_gen: 1, \n",
      " \t number_of_generations: 50000\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-0d3f372c9d14>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    184\u001b[0m             \u001b[0mstep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0man_elementary_act_of_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNe\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpercent_of_recombinating_sp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"nuc\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"first\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m             \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumber_of_generations\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 186\u001b[0;31m                 \u001b[0mstep\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0man_elementary_act_of_evolution\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNe\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpercent_of_recombinating_sp\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"nuc\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mstep\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    187\u001b[0m             \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclock\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-0d3f372c9d14>\u001b[0m in \u001b[0;36man_elementary_act_of_evolution\u001b[0;34m(Ne, seq_len, percent_of_recombinating_sp, mutations_per_gen, nuc_mode, population_mode)\u001b[0m\n\u001b[1;32m    140\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmultiply_seqs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mselected_list\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"crossingover_it\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmultiply_seqs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpopulation\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m     \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mNe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-0d3f372c9d14>\u001b[0m in \u001b[0;36mmultiply_seqs\u001b[0;34m(result, work_list, mutations_per_gen, mode)\u001b[0m\n\u001b[1;32m    135\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    136\u001b[0m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m                     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    138\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    139\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-8-0d3f372c9d14>\u001b[0m in \u001b[0;36mmutate\u001b[0;34m(seq, mutations_per_gen, mode)\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0;31m# creating mutation function -- ok\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmutate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmutations_per_gen\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m         \u001b[0mseq_cp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"new_population\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#%%notify\n",
    "\n",
    "\n",
    "def status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations):\n",
    "    print(\"Recieved parameters: \\n \\t Ne: {Ne}, \\n \\t seq_len: {seq_len}, \\n \\t percent_of_recombinating_sp: {percent_of_recombinating_sp}, \\n \\t mutations_per_gen: {mutations_per_gen}, \\n \\t number_of_generations: {number_of_generations}\".format(Ne = Ne,\n",
    "seq_len = seq_len,\n",
    "percent_of_recombinating_sp = percent_of_recombinating_sp,\n",
    "mutations_per_gen = mutations_per_gen,\n",
    "number_of_generations = number_of_generations))\n",
    "\n",
    "#status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)  \n",
    "\n",
    "\n",
    "\n",
    "# creating a model\n",
    "\n",
    "def an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,nuc_mode=None,population_mode=None):\n",
    "    random.seed(666)\n",
    "    #status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)\n",
    "    \n",
    "    if nuc_mode == \"nuc\":  # выбираем, с чем работаем\n",
    "        seed = \"ATGC\"\n",
    "    else:\n",
    "        seed = \"1234\"\n",
    "    \n",
    "    # creating mutation function -- ok\n",
    "    def mutate(seq, mutations_per_gen, mode=None):\n",
    "        seq_cp = seq.copy()\n",
    "\n",
    "        if mode == \"new_population\":\n",
    "            mutations_per_gen_tmp = mutations_per_gen + int(0.01 * seq_len)\n",
    "            for i in range(mutations_per_gen_tmp):\n",
    "                mutation_site = randrange(len(seq_cp))\n",
    "                seq_cp[mutation_site] = random.choice(seed)     \n",
    "            return seq_cp \n",
    "        else:\n",
    "            r = random.uniform(0, 1) # чтобы было меньше мутаций\n",
    "            if r > 1/10:\n",
    "                return seq_cp\n",
    "            else:\n",
    "                mutations_per_gen_tmp = mutations_per_gen\n",
    "                for i in range(mutations_per_gen_tmp):\n",
    "                    mutation_site = randrange(len(seq_cp))\n",
    "                    seq_cp[mutation_site] = random.choice(seed)     \n",
    "                return seq_cp\n",
    "         \n",
    "    if population_mode == \"first\":   # --ok\n",
    "        # creating seqs\n",
    "        seq = [random.choice(seed) for i in range(seq_len)]\n",
    "        population = [mutate(seq, mutations_per_gen, \"new_population\") for x in range(Ne)]\n",
    "    else:\n",
    "        population = population_mode\n",
    "        Ne = len(population)\n",
    "    # breeding\n",
    "    selected_list = []\n",
    "    number_of_species = int(percent_of_recombinating_sp * len(population))\n",
    "    if number_of_species%2 != 0:\n",
    "        number_of_species -= 1\n",
    "        \n",
    "    for i in range(number_of_species): # --ok\n",
    "        p = randrange(len(population))\n",
    "        selected_list.append(population.pop(p))\n",
    "    \n",
    "    # crossingover\n",
    "    def crossingover(seq_in1, seq_in2):\n",
    "        number_of_parts = random.randrange(0, 2, 1)\n",
    "        if number_of_parts > 0:   \n",
    "            shuffled_1 = []\n",
    "            shuffled_2 = []\n",
    "            left = 0\n",
    "            right = len(seq_in1)\n",
    "            ind = True\n",
    "            for i in range(number_of_parts):\n",
    "                n = number_of_parts - i\n",
    "                #print(right - left)\n",
    "                if right - left > n + 1:\n",
    "                    #print(i, left, right, ind)\n",
    "                    \n",
    "                    #print(\"right//n\", right//n)\n",
    "                    position = random.randint(left, right//n)\n",
    "                    #print(\"position\", position)\n",
    "                    if ind == True:\n",
    "                        shuffled_1.append(seq_in1[left:position])\n",
    "                        shuffled_2.append(seq_in2[left:position])\n",
    "                        #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "                        ind = False\n",
    "                        left = position\n",
    "                        continue\n",
    "                    if ind == False:\n",
    "                        shuffled_1.append(seq_in2[left:position])\n",
    "                        shuffled_2.append(seq_in1[left:position])\n",
    "                        #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "                        ind = True\n",
    "                        left = position\n",
    "                        continue\n",
    "                else:\n",
    "                    #print(\"out\")\n",
    "                    if ind == True:\n",
    "                        shuffled_1.append(seq_in1[left:position])\n",
    "                        shuffled_2.append(seq_in2[left:position])\n",
    "                        #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "                    if ind == False:\n",
    "                        shuffled_1.append(seq_in2[left:position])\n",
    "                        shuffled_2.append(seq_in1[left:position])\n",
    "                        #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "                    return shuffled_1, shuffled_2                        \n",
    "            if ind == True:\n",
    "                shuffled_1.append(seq_in1[left:])\n",
    "                shuffled_2.append(seq_in2[left:])\n",
    "                #print(1, shuffled_1, \"---\", shuffled_2)\n",
    "            if ind == False:\n",
    "                shuffled_1.append(seq_in2[left:])\n",
    "                shuffled_2.append(seq_in1[left:])\n",
    "                #print(2, shuffled_1, \"---\", shuffled_2)\n",
    "            return shuffled_1, shuffled_2\n",
    "        else:\n",
    "            return seq_in1, seq_in2\n",
    "\n",
    "    \n",
    "    \n",
    "    # multipy selected seqs\n",
    "    def multiply_seqs(result, work_list, mutations_per_gen, mode=None):\n",
    "        for i, j in zip(work_list[0::2], work_list[1::2]):\n",
    "            if mode == \"crossingover_it\":\n",
    "                \n",
    "                p, q = crossingover(i, j)\n",
    "                \n",
    "                p = list(itertools.chain.from_iterable(p))\n",
    "                q = list(itertools.chain.from_iterable(q))\n",
    "\n",
    "                for k in range(1): \n",
    "                    result.append(mutate(p, mutations_per_gen))\n",
    "                    result.append(mutate(q, mutations_per_gen))\n",
    "            else:\n",
    "                for k in range(1):\n",
    "                    result.append(mutate(i, mutations_per_gen))\n",
    "                    result.append(mutate(j, mutations_per_gen))\n",
    "        return result\n",
    "    \n",
    "    result = []\n",
    "    result = multiply_seqs(result, selected_list, mutations_per_gen, \"crossingover_it\")\n",
    "    result = multiply_seqs(result, population, mutations_per_gen)\n",
    "    \n",
    "    result = random.sample(result, Ne)\n",
    "    return result\n",
    "\n",
    "\n",
    "\n",
    "def decoder(result):\n",
    "    def make_it_in_nucl(seq):\n",
    "        dic = {1:\"A\",2:\"T\",3:\"G\",4:\"C\"}\n",
    "        for j in range(len(seq)):\n",
    "            seq[j] = dic[int(seq[j])]\n",
    "        return seq\n",
    "    for i in range(len(result)):\n",
    "        result[i] = make_it_in_nucl(result[i])\n",
    "    return result\n",
    "\n",
    "\n",
    "root = \"/home/pavel/Desktop/Work/Project/Simulation\"\n",
    "outputdir = os.path.abspath(root + '/result_0.01')\n",
    "    \n",
    "if not os.path.exists(outputdir):\n",
    "    os.makedirs(outputdir)\n",
    "\n",
    "    \n",
    "# loading parameters\n",
    "mutations_per_gen = 1\n",
    "seq_len = 1000\n",
    "\n",
    "\n",
    "os.chdir(outputdir)\n",
    "for Ne in [x * 5000 for x in range(1, 10, 2)]: #+ [x * 1000 for x in range(2, 10, 2)]:\n",
    "    print(\"--Ne:\", Ne)\n",
    "    for number_of_generations in [x * 10000 for x in range(1, 10, 2)]:# + [x * 1000 for x in range(2, 10, 2)]:\n",
    "        print(\"--number_of_generations:\", number_of_generations)\n",
    "        for percent_of_recombinating_sp in [x * 0.001 for x in range(1, 10, 2)] + [x * 0.01 for x in range(1, 10, 2)] + [x * 0.01 for x in range(1, 10, 2)] + [x * 0.1 for x in range(1, 10, 2)]:\n",
    "            print(\"--percent_of_recombinating_sp:\", percent_of_recombinating_sp)\n",
    "            print(\" \")\n",
    "            \n",
    "            status(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,number_of_generations)  \n",
    "            \n",
    "            start = clock()\n",
    "            step = an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,\"nuc\",\"first\")\n",
    "            for i in range(number_of_generations):\n",
    "                step = an_elementary_act_of_evolution(Ne,seq_len,percent_of_recombinating_sp,mutations_per_gen,\"nuc\",step)\n",
    "            end = clock()\n",
    "            \n",
    "            step = random.sample(step, 100) \n",
    "            \n",
    "        \n",
    "            def count_dict(mystring):\n",
    "                d = {}\n",
    "                for w in mystring: \n",
    "                    d[w] = mystring.count(w)\n",
    "                #print(d.values(), min(d.values())/len(mystring))\n",
    "                if min(d.values())/len(mystring) < 0.25:\n",
    "                    return -1\n",
    "                else:\n",
    "                    return 1\n",
    "\n",
    "            def slice_maker(step):\n",
    "                for i in range(len(step[0])):\n",
    "                    string = \"\"\n",
    "                    for j in range(len(step)):\n",
    "                        string += step[j][i]\n",
    "                    #print(string)\n",
    "                    if count_dict(string) == -1:\n",
    "                        for j in range(len(step)):\n",
    "                            step[j][i] = \"N\"\n",
    "                    \n",
    "            slice_maker(step)\n",
    "\n",
    "            #step = decoder(step)\n",
    "            #print(step)\n",
    "            \n",
    "            \n",
    "            align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
    "            for i in range(len(step)):\n",
    "                align.add_sequence(str(i), \"\".join(step[i]))\n",
    "            \n",
    "            SeqIO.write(align, outputdir + \"/\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \"_align.fasta\", \"fasta\")\n",
    "        \n",
    "            print(\"timing is\", round(end - start,5), \"seconds\")\n",
    "            \n",
    "            try:\n",
    "                line1 = 'snp-sites -mvp -o {out_file} {in_file}'.format(out_file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len),\n",
    "                    in_file = outputdir + \"/\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \"_align.fasta\")\n",
    "                p = subprocess.Popen(line1, shell=True)\n",
    "                p.wait()\n",
    "            \n",
    "                line2 = 'vcftools --hap-r2 --vcf {in_file} --out {out_file}'.format(in_file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".vcf\", \n",
    "                        out_file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len))\n",
    "                p = subprocess.Popen(line2, shell=True)\n",
    "                p.wait()\n",
    "                \n",
    "                    \n",
    "                df = pd.read_csv(outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".hap.ld\", sep='\\t',index_col=False)\n",
    "                df[\"Len\"] = df[\"POS2\"] - df[\"POS1\"]\n",
    "                \n",
    "                df_name = outputdir + \"/LD_df_\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".csv\"\n",
    "                df.to_csv(path_or_buf = df_name)\n",
    "                \n",
    "                line3 = 'Rscript \"/home/pavel/Desktop/Work/Project/Simulation/R_plot.R\" --no-save --no-restore --args {outputdir} {plot_name} 0.0 {df_name}'.format(outputdir = outputdir, \n",
    "                plot_name = outputdir + \"/Ne-\" + str(Ne) + \"_gen_sp-\" + str(number_of_generations) + \"_rec-\" + str(percent_of_recombinating_sp) + \"_mut-\" + str(mutations_per_gen) + \"_seq_l-\" + str(seq_len),\n",
    "                df_name = df_name)\n",
    "                p = subprocess.Popen(line3, shell=True)\n",
    "                p.wait()\n",
    "                \n",
    "                            \n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".phylip\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".snp_sites.aln\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/snp-sites_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".vcf\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len))\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "                    \n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".log\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "\n",
    "                line = 'rm {file}'.format(file = outputdir + \"/vcftools_out_file\" + str(number_of_generations) + \"_\" + str(Ne) + \"_\" + str(percent_of_recombinating_sp) + \"_\" + str(mutations_per_gen) + \"_\" + str(seq_len) + \".hap.ld\")\n",
    "                p = subprocess.Popen(line, shell=True)\n",
    "                    \n",
    "                    \n",
    "            except ZeroDivisionError:\n",
    "                continue\n",
    "        plt.close('all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
